File: Hams alerts.py
#!/usr/bin/env python3
"""
HAMS.AT Alert Checker - NEW ALERTS ONLY
Sends desktop notifications when new satellite alerts appear on hams.at
"""

import requests
import json
import os
from datetime import datetime
from pathlib import Path

# Configuration
API_KEY = "PASTE YOUR API KEY HERE"
API_URL = "https://hams.at/api/alerts/upcoming"
SEEN_ALERTS_FILE = os.path.expanduser("~/.hams_seen_alerts.json")

# Set to True to reset and see all alerts again
RESET_MODE = False


def fetch_alerts():
    """Fetch upcoming alerts from hams.at API"""
    try:
        headers = {
            "Authorization": f"Bearer {API_KEY}"
        }
        response = requests.get(API_URL, headers=headers, timeout=10)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error fetching alerts: {e}")
        return None


def get_seen_alerts():
    """Get list of alert IDs we've already notified about"""
    if RESET_MODE:
        print("RESET MODE: Clearing all seen alerts")
        return []
    
    if os.path.exists(SEEN_ALERTS_FILE):
        try:
            with open(SEEN_ALERTS_FILE, 'r') as f:
                return json.load(f)
        except json.JSONDecodeError:
            return []
    return []


def save_seen_alerts(seen_ids):
    """Save seen alert IDs to file"""
    with open(SEEN_ALERTS_FILE, 'w') as f:
        json.dump(seen_ids, f)


def clean_old_seen_alerts(seen_ids, current_alerts):
    """Keep only IDs that are still in current alerts"""
    current_ids = [alert['id'] for alert in current_alerts]
    return [alert_id for alert_id in seen_ids if alert_id in current_ids]


def format_time(iso_time):
    """Format ISO timestamp to readable format"""
    try:
        dt = datetime.fromisoformat(iso_time.replace('Z', '+00:00'))
        return dt.strftime("%b %d %H:%M")
    except:
        return ""


def send_notification(title, message):
    """Send desktop notification (cross-platform)"""
    try:
        # macOS
        if os.system("which osascript > /dev/null 2>&1") == 0:
            escaped_title = title.replace('"', '\\"')
            escaped_message = message.replace('"', '\\"')
            os.system(f'''osascript -e 'display notification "{escaped_message}" with title "{escaped_title}" sound name "default"' ''')
        
        # Linux (requires notify-send)
        elif os.system("which notify-send > /dev/null 2>&1") == 0:
            os.system(f'notify-send "{title}" "{message}"')
        
        # Windows (requires win10toast - pip install win10toast)
        else:
            try:
                from win10toast import ToastNotifier
                toaster = ToastNotifier()
                toaster.show_toast(title, message, duration=10)
            except ImportError:
                print(f"Notification: {title}\n{message}")
    except Exception as e:
        print(f"Error sending notification: {e}")
        print(f"Notification: {title}\n{message}")


def notify_new_alerts(new_alerts):
    """Send notification about new alerts"""
    if len(new_alerts) == 0:
        print("No new alerts")
        return
    
    title = f"🆕 {len(new_alerts)} New Satellite Alert{'s' if len(new_alerts) > 1 else ''}!"
    
    message_lines = []
    for i, alert in enumerate(new_alerts[:5]):
        callsign = alert.get('callsign', 'Unknown')
        satellite = alert.get('satellite', {}).get('name', '')
        grids = ', '.join(alert.get('grids', []))
        mode = alert.get('mode', '')
        frequency = f"{alert['mhz']} MHz" if alert.get('mhz') else ''
        aos_time = format_time(alert.get('aos_at', ''))
        
        line = f"{callsign}"
        if satellite:
            line += f" • {satellite}"
        if grids:
            line += f" • {grids}"
        
        details = []
        if aos_time:
            details.append(aos_time)
        if mode:
            details.append(mode)
        if frequency:
            details.append(frequency)
        
        if details:
            line += f"\n  {' '.join(details)}"
        
        message_lines.append(line)
    
    if len(new_alerts) > 5:
        message_lines.append(f"\n...and {len(new_alerts) - 5} more")
    
    message = '\n'.join(message_lines)
    
    send_notification(title, message)
    print(f"Notified about {len(new_alerts)} new alerts")


def main():
    """Main execution"""
    data = fetch_alerts()
    
    if data and 'data' in data:
        current_alerts = data['data']
        seen_ids = get_seen_alerts()
        
        # Find new alerts (ones we haven't seen before)
        new_alerts = [alert for alert in current_alerts if alert['id'] not in seen_ids]
        
        if new_alerts:
            # Notify about new alerts
            notify_new_alerts(new_alerts)
            
            # Add new alert IDs to seen list
            new_ids = [alert['id'] for alert in new_alerts]
            seen_ids.extend(new_ids)
        else:
            print("No new alerts found")
        
        # Clean up old alerts from seen list
        seen_ids = clean_old_seen_alerts(seen_ids, current_alerts)
        
        # Save updated seen list
        save_seen_alerts(seen_ids)
        
        print(f"Total alerts: {len(current_alerts)}, New: {len(new_alerts)}, Tracking: {len(seen_ids)}")
    else:
        print("Failed to fetch alerts")


if __name__ == "__main__":
    main()

File: hams_alerts.js
// HAMS.AT Alert Checker - NEW ALERTS ONLY
// Only sends notifications when new alerts appear

const API_KEY = "PASTE YOUR API KEY HERE";
const API_URL = "https://hams.at/api/alerts/upcoming";
const SEEN_ALERTS_KEY = "hams_seen_alerts";

// SET THIS TO true TO RESET AND SEE ALL ALERTS AGAIN
const RESET_MODE = false;

async function fetchAlerts() {
  try {
    let req = new Request(API_URL);
    req.headers = {
      "Authorization": `Bearer ${API_KEY}`
    };
    
    let response = await req.loadJSON();
    return response;
  } catch (error) {
    console.error("Error fetching alerts: " + error);
    return null;
  }
}

function getSeenAlerts() {
  if (RESET_MODE) {
    console.log("RESET MODE: Clearing all seen alerts");
    return [];
  }
  
  // Get list of alert IDs we've already notified about
  let fm = FileManager.iCloud();
  let path = fm.joinPath(fm.documentsDirectory(), SEEN_ALERTS_KEY);
  
  if (fm.fileExists(path)) {
    let data = fm.readString(path);
    return JSON.parse(data);
  }
  return [];
}

function saveSeenAlerts(seenIds) {
  let fm = FileManager.iCloud();
  let path = fm.joinPath(fm.documentsDirectory(), SEEN_ALERTS_KEY);
  fm.writeString(path, JSON.stringify(seenIds));
}

function cleanOldSeenAlerts(seenIds, currentAlerts) {
  // Keep only IDs that are still in current alerts
  // This prevents the list from growing forever
  let currentIds = currentAlerts.map(a => a.id);
  return seenIds.filter(id => currentIds.includes(id));
}

async function notifyNewAlerts(newAlerts) {
  if (newAlerts.length === 0) {
    console.log("No new alerts");
    return;
  }
  
  let notification = new Notification();
  notification.title = `🆕 ${newAlerts.length} New Satellite Alert${newAlerts.length > 1 ? 's' : ''}!`;
  
  let body = "";
  for (let i = 0; i < Math.min(newAlerts.length, 5); i++) {
    let alert = newAlerts[i];
    
    let callsign = alert.callsign || "Unknown";
    let satellite = alert.satellite ? alert.satellite.name : "";
    let grids = alert.grids && alert.grids.length > 0 ? alert.grids.join(", ") : "";
    let mode = alert.mode || "";
    let frequency = alert.mhz ? `${alert.mhz} MHz` : "";
    
    // Format time
    let aosTime = "";
    if (alert.aos_at) {
      let date = new Date(alert.aos_at);
      let dateFormatter = new DateFormatter();
      dateFormatter.dateFormat = "MMM d HH:mm";
      aosTime = dateFormatter.string(date);
    }
    
    body += `${callsign}`;
    if (satellite) body += ` • ${satellite}`;
    if (grids) body += ` • ${grids}`;
    body += "\n";
    if (aosTime) body += `  ${aosTime}`;
    if (mode) body += ` ${mode}`;
    if (frequency) body += ` ${frequency}`;
    body += "\n";
  }
  
  if (newAlerts.length > 5) {
    body += `\n...and ${newAlerts.length - 5} more`;
  }
  
  notification.body = body.trim();
  notification.sound = "default";
  
  await notification.schedule();
  console.log(`Notified about ${newAlerts.length} new alerts`);
}

// Main execution
let data = await fetchAlerts();

if (data && data.data) {
  let currentAlerts = data.data;
  let seenIds = getSeenAlerts();
  
  // Find new alerts (ones we haven't seen before)
  let newAlerts = currentAlerts.filter(alert => !seenIds.includes(alert.id));
  
  if (newAlerts.length > 0) {
    // Notify about new alerts
    await notifyNewAlerts(newAlerts);
    
    // Add new alert IDs to seen list
    let newIds = newAlerts.map(a => a.id);
    seenIds = seenIds.concat(newIds);
  } else {
    console.log("No new alerts found");
  }
  
  // Clean up old alerts from seen list
  seenIds = cleanOldSeenAlerts(seenIds, currentAlerts);
  
  // Save updated seen list
  saveSeenAlerts(seenIds);
  
  console.log(`Total alerts: ${currentAlerts.length}, New: ${newAlerts.length}, Tracking: ${seenIds.length}`);
  
} else {
  console.error("Failed to fetch alerts");
}

// Widget display
if (config.runsInWidget) {
  let widget = new ListWidget();
  widget.backgroundColor = new Color("#1c1c1e");
  
  let title = widget.addText("🛰️ HAMS.AT");
  title.textColor = Color.white();
  title.font = Font.boldSystemFont(16);
  
  widget.addSpacer(8);
  
  if (data && data.data) {
    let count = widget.addText(`${data.data.length} upcoming alerts`);
    count.textColor = Color.orange();
    count.font = Font.systemFont(14);
  } else {
    let error = widget.addText("Error loading");
    error.textColor = Color.red();
    error.font = Font.systemFont(14);
  }
  
  Script.setWidget(widget);
}

Script.complete();
